system_tests/recreatestate_rpc_test.go

```diff
@@ -259,9 +262,166 @@ func TestRecreateStateForRPCBlockNotFoundWhileRecreating(t *testing.T) {
 	if err == nil {
 		hash := rawdb.ReadCanonicalHash(db, lastBlock)
-		testhelpers.FailImpl(t, "Didn't fail to get balance at block:", lastBlock, " with hash:", hash, ", lastBlock:", lastBlock)
+		Fatal(t, "Didn't fail to get balance at block:", lastBlock, " with hash:", hash, ", lastBlock:", lastBlock)
 	}
 	if !strings.Contains(err.Error(), "block not found while recreating") {
-		testhelpers.FailImpl(t, "Failed with unexpected error: \"", err, "\", at block:", lastBlock, "lastBlock:", lastBlock)
+		Fatal(t, "Failed with unexpected error: \"", err, "\", at block:", lastBlock, "lastBlock:", lastBlock)
+	}
+}
+
+func testSkippingSavingStateAndRecreatingAfterRestart(t *testing.T, skipBlocks uint32, skipGas uint64, txCount int) {
+	t.Helper()
+	maxRecreateStateDepth := int64(30 * 1000 * 1000)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ctx1, cancel1 := context.WithCancel(ctx)
+	nodeConfig := arbnode.ConfigDefaultL2Test()
+	nodeConfig.RPC.MaxRecreateStateDepth = maxRecreateStateDepth
+	nodeConfig.Sequencer.MaxBlockSpeed = 0
+	nodeConfig.Sequencer.MaxTxDataSize = 150 // 1 test tx ~= 110
+	cacheConfig := &core.CacheConfig{
+		// Arbitrum Config Options
+		TriesInMemory:                      128,
+		TrieRetention:                      30 * time.Minute,
+		MaxNumberOfBlocksToSkipStateSaving: skipBlocks,
+		MaxAmountOfGasToSkipStateSaving:    skipGas,
+
+		// disable caching of states in BlockChain.stateCache
+		TrieCleanLimit: 0,
+		TrieDirtyLimit: 0,
+
+		TrieDirtyDisabled: true,
+
+		TrieTimeLimit: 5 * time.Minute,
+		SnapshotLimit: 256,
+		SnapshotWait:  true,
+	}
+
+	feedErrChan := make(chan error, 10)
+	AddDefaultValNode(t, ctx1, nodeConfig, true)
+	l2info, stack, chainDb, arbDb, blockchain := createL2BlockChain(t, nil, t.TempDir(), params.ArbitrumDevTestChainConfig(), cacheConfig)
+
+	node, err := arbnode.CreateNode(ctx1, stack, chainDb, arbDb, NewFetcherFromConfig(nodeConfig), blockchain, nil, nil, nil, nil, nil, feedErrChan)
+	Require(t, err)
+	err = node.TxStreamer.AddFakeInitMessage()
+	Require(t, err)
+	Require(t, node.Start(ctx1))
+	client := ClientForStack(t, stack)
+	debugAuth := l2info.GetDefaultTransactOpts("Owner", ctx1)
+	// make auth a chain owner
+	arbdebug, err := precompilesgen.NewArbDebug(common.HexToAddress("0xff"), client)
+	Require(t, err, "failed to deploy ArbDebug")
+	tx, err := arbdebug.BecomeChainOwner(&debugAuth)
+	Require(t, err, "failed to deploy ArbDebug")
+	_, err = EnsureTxSucceeded(ctx1, client, tx)
+	Require(t, err)
+
+	StartWatchChanErr(t, ctx, feedErrChan, node)
+	dataDir := node.Stack.DataDir()
+
+	l2info.GenerateAccount("User2")
+	var txs []*types.Transaction
+	for i := 0; i < txCount; i++ {
+		tx := l2info.PrepareTx("Owner", "User2", l2info.TransferGas, common.Big1, nil)
+		txs = append(txs, tx)
+		err := client.SendTransaction(ctx, tx)
+		Require(t, err)
+	}
+	for _, tx := range txs {
+		_, err := EnsureTxSucceeded(ctx, client, tx)
```

Comment:
we rely heavily on which tx is in which block in this test. We can check it here.
EnsureTxSucceeded returns a receipt where you can check block number and fail if it's not what's expected, which will make debugging it easier.